_posts/2019-08-23-unet/unet.Rmd

```diff
@@ -0,0 +1,437 @@
+
+The network architecture used in this post is adequate for _class segmentation_ tasks and should be applicable to a vast number of practical, scientific as well as non-scientific applications. Speaking of network architecture, how should it look?
+
+## Introducing U-Net
+
+Given their success in image classification, can't we just use a classic architecture like _Inception V[n]_, _ResNet_, _ResNext_ ... , whatever? The problem is, our task at hand -- labeling every pixel -- does not fit so well with the classic idea of a CNN. With convnets, the idea is to apply successive layers of convolution and pooling to build up feature maps of decreasing granularity, to finally arrive at an abstract level where we just say: "yep, a cat". The other side of the medail is, we lose detail information: To the final classification, it does not matter whether the five pixels in the top-left area are black or white.
+
+In practice, the classic architectures use (max) pooling or convolutions with `stride` > 1 to achieve those successive abstractions -- necessarily resulting in decreased spatial resolution.
+So how can we use a convnet and still preserve detail information? In their 2015 paper _U-Net: Convolutional Networks for Biomedical Image Segmentation_ [@RonnebergerFB15], Olaf Ronneberger et al. came up with what four years later, in 2019, is still the most popular approach. (Which is to say something, four years being a long time, in deep learning.)
+
+The idea is stunningly simple. While successive encoding (convolution / max pooling) steps, as usual, reduce resolution, the subsequent decoding -- we have to arrive at an output of  size same as the input, as we want to label every pixel! -- does not simply upsample from the most compressed layer. Instead, during upsampling, at every step we feed in information from the corresponding, in resolution, layer in the downsizing chain.
+
+For U-Net, really a picture says more than many words:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "U-Net architecture from Ronneberger et al. 2015."}
+knitr::include_graphics("images/unet.png")
+```
+
+
+Specifically, at each upsampling stage we _concatenate_ the output from the previous layer with that from its pendant in the compression stage. The final output is a _mask_ of size the original image, obtained via 1x1-convolution; no final dense layer is required, instead the output layer is just a convolutional layer with a single filter. 
```

Comment:
```suggestion
At each upsampling stage we _concatenate_ the output from the previous layer with that from its pendant in the compression stage. The final output is a _mask_ of size the original image, obtained via 1x1-convolution; no final dense layer is required, instead the output layer is just a convolutional layer with a single filter. 
```

Reply:
I also don't know what a pendant is, but I don't know very much about this topic.

Reply:
lol again! placeholder for sth to look up in the dictionary ... totally overlooked when proofreading ... wait ...
... that  will be "counterpart" :-)))